src/common.rs

```diff
@@ -333,6 +333,67 @@ impl Default for Compression {
     }
 }
 
+/// Advanced compression settings with more customization options than [Compression].
+///
+/// Note that this setting only affects DEFLATE compression.
+/// Another setting that influences the compression ratio and lets you choose
+/// between encoding speed and compression ratio is the *filter*,
+/// See [FilterType] and [AdaptiveFilterType].
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy)]
+pub enum DeflateCompression {
+    /// Do not compress the data at all.
+    ///
+    /// Useful for incompressible images such as photographs,
+    /// or when speed is paramount and you don't care about size at all.
+    ///
+    /// This mode also disables filters, forcing [FilterType::NoFilter].
+    NoCompression,
+
+    /// Excellent for creating lightly compressed PNG images very quickly.
+    ///
+    /// Uses the [fdeflate](https://crates.io/crates/fdeflate) crate under the hood
+    /// to achieve speeds far exceeding what libpng is capable of
+    /// while still providing a decent compression ratio.
+    ///
+    /// Images encoded in this mode can also be decoded by the `png` crate extremely quickly,
+    /// much faster than what is typical for PNG images.
+    /// Other decoders (e.g. libpng) do not get a decoding speed boost from this mode.
+    FdeflateUltraFast,
+
+    /// Uses [flate2](https://crates.io/crates/flate2) crate with the specified [compression level](flate2::Compression::new).
+    ///
+    /// Flate2 has several backends that make different trade-offs.
+    /// See the flate2 documentation for the available backends for more information.
+    Flate2(u32),
+    // TODO: Zopfli?
```

Comment:
Should this be dependent on feature flags?

If they exist unconditionally, then they can either fail at run time if a dependency is not added, or we'll have to have all the deps included all the time.

Reply:
All the modes listed right now are unconditional dependencies.

I'm really not a fan of things failing at runtime due to improper build time configuration, so for Zopfli I was going to gate that enum variant behind a feature. That way we could let the API user decide whether want it and enable the feature, decide they don't want it and disable the feature, or pass on the optionality of this by exposing their own feature, which would affect this enum.

I'm happy to hear your thoughts on how an optional Zopfli dependency could be added in the future, and how can we adjust this API to make that easier.

Reply:
There are even nice rustdoc annotations to make it clear that something is only available when a certain feature is enabled.

Reply:
I'd prefer to ship this with a zopfli Cargo feature, even if it's default on, even if it doesn't actually disable Zopfli yet, only the config option. 

This is because addition of a feature flag would be a breaking change, but after that the change whether the flag actually removes zopfli dependency isn't breaking.

Reply:
The enum is marked `#[non_exhaustive]` so the addition of new variants is not a breaking change.

Reply:
But a conditional removal of `Zopfli` variant via Cargo feature is a breaking change.

It would have to be completely removed before release in order to make it backwards-compatible to add it behind a flag.

Reply:
Just so we're on the same page: I have no plans to add a Zopfli variant in this PR. If/when a Zopfli variant arrives, it will be gated on a Cargo feature from the start.

Reply:
I've removed the mentions of Zopfli from this PR entirely.